overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,466 @@
+	// TODO: change to this once PR #13730 is merged:
+	// return PathPatternMatch(c.PathPattern, path)
+	return true, nil
+}
+
+// RemovePermission removes every instance of the given permission from the
+// permissions list associated with the constraints. If the permission does
+// not exist in the list, returns ErrPermissionNotInList.
+func (c *Constraints) RemovePermission(permission string) error {
+	origLen := len(c.Permissions)
+	for i := 0; i < len(c.Permissions); {
+		perm := c.Permissions[i]
+		if perm != permission {
+			i++
+			continue
+		}
+		copy(c.Permissions[i:], c.Permissions[i+1:])
+		c.Permissions = c.Permissions[:len(c.Permissions)-1]
+	}
+	if origLen == len(c.Permissions) {
+		return ErrPermissionNotInList
+	}
+	return nil
+}
+
+// ContainPermissions returns true if the constraints include every one of the
+// givne permissions.
+func (c *Constraints) ContainPermissions(permissions []string) bool {
+	for _, perm := range permissions {
+		if !strutil.ListContains(c.Permissions, perm) {
+			return false
+		}
+	}
+	return true
+}
+
+type OutcomeType string
+
+const (
+	OutcomeUnset OutcomeType = ""
+	OutcomeAllow OutcomeType = "allow"
+	OutcomeDeny  OutcomeType = "deny"
+)
+
+// AsBool returns the outcome as a boolean, or an error if it cannot be parsed.
+func (outcome OutcomeType) AsBool() (bool, error) {
+	switch outcome {
+	case OutcomeAllow:
+		return true, nil
+	case OutcomeDeny:
+		return false, nil
+	default:
+		return false, fmt.Errorf(`invalid outcome: must be %q or %q: %q`, OutcomeAllow, OutcomeDeny, outcome)
+	}
+}
+
+type LifespanType string
+
+const (
+	LifespanUnset    LifespanType = ""
+	LifespanForever  LifespanType = "forever"
+	LifespanSession  LifespanType = "session"
+	LifespanSingle   LifespanType = "single"
+	LifespanTimespan LifespanType = "timespan"
+)
+
+// TimestampToTime converts the given timestamp string to a time.Time in Local
+// time. The timestamp string is expected to be of the format time.RFC3339Nano.
+// If it cannot be parsed as such, returns an error.
+func TimestampToTime(timestamp string) (time.Time, error) {
+	t, err := time.Parse(time.RFC3339Nano, timestamp)
+	if err != nil {
+		return t, err
+	}
+	return t.Local(), nil
+}
+
+// CurrentTimestamp returns the current time as a time.RFC3339Nano string.
+func CurrentTimestamp() string {
+	return time.Now().Format(time.RFC3339Nano)
+}
+
+// NewID returns a new unique ID.
+//
+// The ID is the current unix time in nanoseconds encoded as base32.
+func NewID() string {
+	nowUnix := uint64(time.Now().UnixNano())
+	nowBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(nowBytes, nowUnix)
+	id := base32.StdEncoding.EncodeToString(nowBytes)
+	return id
+}
+
+// NewIDAndTimestamp returns a new unique ID and corresponding timestamp.
+//
+// The ID is the current unix time in nanoseconds encoded as a string in base32.
+// The timestamp is the same time, encoded as a string in time.RFC3339Nano.
+func NewIDAndTimestamp() (id string, timestamp string) {
```

Comment:
this could either go to timeutil or maybe be introduced when they are used? it's a bit uncler if they need to be exported or not